### PR TITLE
Addon Test: Support story name as test description

### PR DESCRIPTION
**Pull request by InfiniteXyy**
## What I did

I added a new feature to the newly published `vitest-test` plugin:

Now the test description will use the explictly settled `name` of the story, and then fallback to the `exportName`.

The requirement comes from that developers might want to have spaces/brackets/etc... in the test description, which is not allowed in exportName.

At the same time, this can also keep consistent with the name display in storybook preview.

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [x] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [x] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.4 MB | 77.4 MB | 0 B | **1.54** | 0% |
| initSize |  162 MB | 162 MB | 409 B | -0.21 | 0% |
| diffSize |  85 MB | 85 MB | 409 B | -0.39 | 0% |
| buildSize |  7.57 MB | 7.57 MB | 0 B | -0.58 | 0% |
| buildSbAddonsSize |  1.66 MB | 1.66 MB | 0 B | -0.58 | 0% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  2.34 MB | 2.34 MB | 0 B | - | 0% |
| buildSbPreviewSize |  352 kB | 352 kB | 0 B | - | 0% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.55 MB | 4.55 MB | 0 B | -0.58 | 0% |
| buildPreviewSize |  3.02 MB | 3.02 MB | 0 B | -0.58 | 0% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  7.2s | 10.4s | 3.1s | -0.46 | 30.6% |
| generateTime |  21.5s | 21.2s | -309ms | 0.8 | -1.5% |
| initTime |  16s | 17s | 1s | 0.64 | 5.9% |
| buildTime |  10.6s | 12.9s | 2.3s | **2.26** | 🔺18.1% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  7.3s | 6.3s | -947ms | -1.03 | -14.9% |
| devManagerResponsive |  5s | 4.3s | -739ms | -0.53 | -17% |
| devManagerHeaderVisible |  709ms | 723ms | 14ms | -0.86 | 1.9% |
| devManagerIndexVisible |  740ms | 754ms | 14ms | -0.9 | 1.9% |
| devStoryVisibleUncached |  1.1s | 1.1s | 7ms | -0.89 | 0.6% |
| devStoryVisible |  738ms | 753ms | 15ms | -0.91 | 2% |
| devAutodocsVisible |  639ms | 645ms | 6ms | -1.19 | 0.9% |
| devMDXVisible |  606ms | 630ms | 24ms | **-1.31** | 3.8% |
| buildManagerHeaderVisible |  644ms | 694ms | 50ms | **-1.25** | 🔺7.2% |
| buildManagerIndexVisible |  649ms | 701ms | 52ms | **-1.32** | 🔺7.4% |
| buildStoryVisible |  735ms | 768ms | 33ms | -1.02 | 4.3% |
| buildAutodocsVisible |  639ms | 604ms | -35ms | **-1.4** | 🔰-5.8% |
| buildMDXVisible |  572ms | 630ms | 58ms | -0.92 | 9.2% |

## Greptile Summary

This PR enhances the Vitest plugin for Storybook by allowing customizable test descriptions using the story's 'name' property, improving readability and consistency with the Storybook preview.

- Modified `code/core/src/csf-tools/vitest-plugin/transformer.ts` to prioritize story 'name' over 'exportName' for test descriptions
- Added new test case in `code/core/src/csf-tools/vitest-plugin/transformer.test.ts` to verify the 'name' property usage
- Updated `docs/writing-tests/vitest-plugin.mdx` with instructions on customizing test descriptions using the 'name' property
- This change allows for more descriptive test names, including spaces and special characters not permitted in export names

### CODE REVIEW

File: code/core/src/csf-tools/vitest-plugin/transformer.ts [hunk 1]
```diff
@@ -207,10 +207,34 @@ export async function vitestTransform({
       exportName: string;
       node: t.Node;
     }): t.ExpressionStatement => {
+      // Get test name from the story's name property, and fallback to exportName
+      const testName = (function getTestName(): string {
```
[the remark this answers is not in the document]
Comment (InfiniteXyy) [reply code/core/src/csf-tools/vitest-plugin/transformer.ts]: remove the usage of iife

File: code/core/src/csf-tools/vitest-plugin/transformer.ts [hunk 2]
```diff
@@ -207,10 +207,34 @@ export async function vitestTransform({
...
+                // Filter out the "name" property and return its value
+                prop.type === 'ObjectProperty' &&
+                prop.key.type === 'Identifier' &&
+                prop.value.type === 'StringLiteral' &&
+                prop.key.name === 'name'
+              ) {
+                return prop.value.value;
+              }
+            }
+          }
```
[the remark this answers is not in the document]
Comment (InfiniteXyy) [reply code/core/src/csf-tools/vitest-plugin/transformer.ts]: With find, the type will be hard to define

**Review by yannbf — CHANGES_REQUESTED** [review]
Hey @InfiniteXyy thanks a lot for your contribution! This was something we were planning to do, so thanks for speeding that up! I'd love if you could change the implemention to make it simpler, given that the underlying parsing tool (CsfTools) already handles extracting the story name.

I believe you can delete most of the code and use something like this:
```ts
const exportName = parsed._stories[exportName].name || exportName;
return getTestStatementForStory({
          exportName,
          node,
});
```

Test cases:

CSF1/2
```ts
export const Story = () => {}
Story.storyName = 'custom name'
```

CSF3
```ts
export const Story = () => {
  name: 'custom name'
}
```

Let me know how it goes and if you need anything else!

File: docs/writing-tests/vitest-plugin.mdx [hunk 3]
```diff
@@ -379,6 +379,10 @@ We recommend running tests in a browser using Playwright, but you can use WebDri
 
 We recommend using Chromium, because it is most likely to best match the experience of a majority of your users. However, you can use other browsers by adjusting the [browser name in the Vitest configuration file](https://vitest.dev/config/#browser-name). Note that [Playwright and WebDriverIO support different browsers](https://vitest.dev/guide/browser/#browser-option-types).
 
+### How do I customized the test description
+
+We are using the exportName of a story definition by default, but you can provide a `name` property for the story to customize the test description. This is useful when you want have places, brackets or other special characters in the test description.
```
Comment (yannbf) [inline docs/writing-tests/vitest-plugin.mdx:384]: Can you add a code snippet alongside this so users can understand it better?

```ts
export const Story = {
  storyName: 'custom, descriptive name'
};
```

```suggestion
### How do I customize the test description?

By default, the export name of a story is mapped to the test name. You can provide a `name` property for the story to customize the test description, allowing you to write more descriptive names. This is useful when you want use spaces, brackets or other special characters in the test description.
```

**Comment by yannbf** [comment]
Also, can you share examples of how you are naming your stories for your tests? It's pretty useful for us to have an idea about use cases! Thanks!

**Comment by InfiniteXyy** [comment]
> Also, can you share examples of how you are naming your stories for your tests? It's pretty useful for us to have an idea about use cases! Thanks!

Hi, thank you for your code review and feed back.
In our project, we are following a naming convention in test descriptions.

Each test cases should have a description like `WHEN user click reply, THEN a request should be sent with a double confirm dialog`.

It's more like a sentense, because we want each test to have a meaningful description that represents a specific scenario.

**Comment by InfiniteXyy** [comment]
ser

> Hey @InfiniteXyy thanks a lot for your contribution! This was something we were planning to do, so thanks for speeding that up! I'd love if you could change the implemention to make it simpler, given that the underlying parsing tool (CsfTools) already handles extracting the story name.
> 
> I believe you can delete most of the code and use something like this:
> 
> ```ts
> const exportName = parsed._stories[exportName].name || exportName;
> return getTestStatementForStory({
>           exportName,
>           node,
> });
> ```
> 
> Test cases:
> 
> CSF1/2
> 
> ```ts
> export const Story = () => {}
> Story.storyName = 'custom name'
> ```
> 
> CSF3
> 
> ```ts
> export const Story = () => {
>   name: 'custom name'
> }
> ```
> 
> Let me know how it goes and if you need anything else!

Thank you for your suggestions! It's helpful and by using `parsedStory`, the code looks just much better now.

I also try adding the test cases for csf v1/2 and function config, but I am not familiar with csf v1/2, hope I didn't do it wrong.

**Comment by yannbf** [comment]
Hey @InfiniteXyy I'm trying to wrap this PR up but I am not allowed to push to your remote. Would you mind [allowing edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) in your fork? Thanks!

**Comment by InfiniteXyy** [comment]
> Hey @InfiniteXyy I'm trying to wrap this PR up but I am not allowed to push to your remote. Would you mind [allowing edits from maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork) in your fork? Thanks!

I have enabled this, hope it's the correct place for allowing permission
image

**Review by yannbf — APPROVED** [review]
Thank you so much for your contribution!

**Pull request merged**